src/plasma/plasma_client.cc

```diff
@@ -465,13 +469,67 @@ void plasma_contains(PlasmaConnection *conn, ObjectID obj_id, int *has_object) {
   }
 }
 
+void compute_block_hash(const unsigned char *data, int64_t nbytes, uint64_t *hash) {
+  XXH64_state_t hash_state;
+  XXH64_reset(&hash_state, XXH64_DEFAULT_SEED);
+  XXH64_update(&hash_state, data, nbytes);
+  *hash = XXH64_digest(&hash_state);
+}
+
+bool compute_object_hash_parallel(XXH64_state_t *hash_state,
+                                  const unsigned char *data,
+                                  int64_t nbytes) {
+#ifndef NUMTHREADS
+#define NUMTHREADS 8
+#endif
+  const uint64_t numthreads = NUMTHREADS;
+  uint64_t threadhash[numthreads+2];
+  //CHECK(numthreads > 0);
+  const uint64_t blocksz = 64; // cache block alignment (alternative: page size)
+  // Calculate the first and last aligned positions in the data stream.
+  unsigned char *databp = (unsigned char *)(((uint64_t)data + blocksz-1) & ~(blocksz-1));
```

Comment:
this is meant to work for blocksz = any power of 2. The only way to simplify is to assume some specific alignment and substitute the constants. Let's assume 64 byte alignment.  The first term on the rhs, tips the pointer over the closest aligned memory position. The second term is the negation of the power of 2 alignment minus 1. So it leaves all bits on except the bits that correspond to the bit-width of the alignment (in this case : 6 least significant bits will be zero). Doing a binary AND of the first term with the second term zeros out 6 least significant bits, thus causing the result to be 64-byte aligned to the next 64-byte aligned memory position. If it's already aligned, it stays the same.